ovf/parser.go

```diff
@@ -0,0 +1,253 @@
+/*
+Copyright (c) 2019-2024 VMware, Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ovf
+
+import (
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// These are used to validate the overall structure of the string being parsed and to differentiate tokens as we are
+// processing them
+var (
+	validIntegerRegexp  = regexp.MustCompile(`^([1-9]\d*)$`)
+	validExponentRegexp = regexp.MustCompile(`^([1-9]\d*\^[1-9]\d*)$`)
+	validByteUnitRegexp = regexp.MustCompile(`((^|kilo|kibi|mega|mebi|giga|gibi)byte(s?)$)`)
+	validCapacityRegexp = regexp.MustCompile(`^[[:blank:]]*((([1-9]\d*\^[1-9]\d*)|([1-9]\d*))($|[[:blank:]]*\*[[:blank:]]*))*(([a-zA-Z]*(b|B)(y|Y)(t|T)(e|E)(s|S)?)($|([[:blank:]]*\*[[:blank:]]*(([1-9]\d*\^[1-9]\d*)|([1-9]\d*)))*))?$`)
+)
+
+// We only handle kilo, kibi, mega, mebi, giga, gibi prefixes due to size constraints of int64/uint64, but more
+// importantly because prefixes larger than giga & gibi don't make sense for our use-case
+var prefixMultipliers = map[string]int64{
+	"byte":     1,                      // byte
+	"kilobyte": 1 * 1000,               // byte * 1000
+	"kibibyte": 1 * 1024,               // byte * 1024
+	"megabyte": 1 * 1000 * 1000,        // byte * 1000 * 1000 = kilobyte * 1000
+	"mebibyte": 1 * 1024 * 1024,        // byte * 1024 * 1024 = kibibyte * 1024
+	"gigabyte": 1 * 1000 * 1000 * 1000, // byte * 1000 * 1000 * 1000 = kilobyte * 1000 * 1000 = megabyte * 1000
+	"gibibyte": 1 * 1024 * 1024 * 1024, // byte * 1024 * 1024 * 1024 = kibibyte * 1024 * 1024 = mebibyte * 1024
+}
+
+// ParseCapacityAllocationUnits validates the string s is a valid programmatic unit with respect to the base unit 'byte'
+// and parses the string to return the number of bytes s represents
+func ParseCapacityAllocationUnits(s string) int64 {
+	var capacityBytes int64
+
+	// Any strings which don't match against the regular expression are deemed invalid and zero is returned as the result
+	if validCapacityString(s) {
```

Comment:
Can you please invert the condition and return early? That way the majority of the logic is dedented and easier to read. Thanks!